### PR TITLE
Change `debug_symbols` SCons argument from bool to enum

### LINKED ISSUE #72247: [Build System] separate_build_symbols doesn't enable build_symbols
### Godot version

master

### System information

Win 11

### Issue description

As the final compiled binary with `debug_symbols=True` is currently around 1.8GB, I wanted to see if I can compile the debug symbols into a separate file (so I can also use the compiled binary, as it slows down significantly when using the 1.8GB one).
Trying to compile with just `separate_debug_symbols=True` doesn't produce anything at all. It only seems to work when also using `debug_symbols=True` - so I think it should automatically enable `debug_symbols` if we set `separate_debug_symbols`.

image

### Steps to reproduce

* get sources
* compile with `scons separate_debug_symbols=true` 
* don't receive debug symbols

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by brno32**
resolves https://github.com/godotengine/godot/issues/72247

Seems simpler than adding logic to coerce `debug_symbols` to `True` when `debug_symbols=yes` is passed

EDIT: switching to an enum while preserving old behavior

### CODE REVIEW

**Comment by Calinou** [comment]
In the long term, we should move this setting to an enum with the values `debug_symbols=no|embedded|separate`.

File: SConstruct [hunk 1]
```diff
@@ -178,8 +178,21 @@ opts.Add(
         "optimize", "Optimization level", "speed_trace", ("none", "custom", "debug", "speed", "speed_trace", "size")
     )
 )
-opts.Add(BoolVariable("debug_symbols", "Build with debugging symbols", False))
-opts.Add(BoolVariable("separate_debug_symbols", "Extract debugging symbols to a separate file", False))
+opts.Add(
+    EnumVariable(
+        "debug_symbols",
+        "Build with debugging symbols (options are no, embedded, and separate, but yes is still supported for backwards compatibility)",
```
Comment (Calinou) [inline SConstruct:184]: The options don't need to be mentioned here, as validation will mention the valid enum values automatically.

File: SConstruct [hunk 2]
```diff
@@ -178,8 +178,21 @@ opts.Add(
...
+        "no",
+        ("no", "yes", "embedded", "separate"),
+    )
+)
+opts.Add(
+    BoolVariable(
+        "separate_debug_symbols",
+        "Extract debugging symbols to a separate file (can also be achieved with debug_symbols=separate)",
```
Comment (Calinou) [inline SConstruct:192]:
```suggestion
        "Extract debugging symbols to a separate file (deprecated, use debug_symbols=separate instead)",
```

**Comment by adamscott** [comment]
> Old values should be read for compatibility with existing scripts, with `yes` being `embedded` and `separate_debug_symbols=yes` turning this into `separate` if `debug_symbols` is `yes`.

Technically, if we want true compatibility, we should accept all the values of `BoolVariable`.
```python
TRUE_STRINGS = ('y', 'yes', 'true', 't', '1', 'on' , 'all')
FALSE_STRINGS = ('n', 'no', 'false', 'f', '0', 'off', 'none')
```

https://scons.org/doc/latest/HTML/scons-api/_modules/SCons/Variables/BoolVariable/

File: SConstruct [hunk 3]
```diff
@@ -178,8 +178,21 @@ opts.Add(
...
     )
 )
-opts.Add(BoolVariable("debug_symbols", "Build with debugging symbols", False))
-opts.Add(BoolVariable("separate_debug_symbols", "Extract debugging symbols to a separate file", False))
+opts.Add(
+    EnumVariable(
+        "debug_symbols",
+        "Build with debugging symbols",
+        "no",
+        ("no", "yes", "embedded", "separate"),
```
Comment (adamscott) [inline SConstruct:186]:
```suggestion
        ("embedded", "separate") + TRUE_STRINGS + FALSE_STRINGS,
```

Comment (akien-mga) [reply SConstruct:186]: The `--help` output would become unreadable :/

File: SConstruct [hunk 4]
```diff
@@ -178,8 +180,26 @@ opts.Add(
...
-opts.Add(BoolVariable("debug_symbols", "Build with debugging symbols", False))
-opts.Add(BoolVariable("separate_debug_symbols", "Extract debugging symbols to a separate file", False))
+legacy_values = TRUE_STRINGS + FALSE_STRINGS
+debug_symbols = EnumVariable(
+    "debug_symbols",
+    "Build with debugging symbols",
+    "no",
+    # values from TRUE_STRINGS and FALSE_STRINGS are supported for backwards compatibility
+    ("no", "embedded", "separate") + legacy_values,
+)
```
Comment (adamscott) [inline SConstruct:190]:
```suggestion
debug_symbols = EnumVariable(
    "debug_symbols",
    "Build with debugging symbols (deprecated value `yes` defaults to `embedded`)",
    "no",
    # values from TRUE_STRINGS and FALSE_STRINGS are supported for backwards compatibility
    ("yes", "no", "embedded", "separate") + legacy_values,
)
```

I would add still `yes` as a "value", but say that it's deprecated.

**Review by Calinou — CHANGES_REQUESTED** [review]
Tested locally (rebased against `master` https://github.com/godotengine/godot/commit/442bbb595037321fd4d4224c7dde151525bb9538), it doesn't work as expected on Linux:

- `debug_symbols=embedded` is ignored and acts like `debug_symbols=no`.
- `debug_symbols=separate` is ignored and acts like `debug_symbols=no`.
- `debug_symbols=yes` works correctly.

Also, we may want to print a message if using `debug_symbols=yes` or `separate_debug_symbols=yes` to tell users to use `debug_symbols=embedded` and `debug_symbols=separate` still.

Like this: https://github.com/godotengine/godot/blob/442bbb595037321fd4d4224c7dde151525bb9538/SConstruct#L325

Full command line I'm using for testing:

```
scons -j32 platform=linuxbsd linker=mold optimize=none fast_unsafe=yes progress=no debug_symbols=embedded builtin_embree=no builtin_enet=no builtin_freetype=no builtin_graphite=no builtin_harfbuzz=no builtin_libogg=no builtin_libpng=no builtin_libtheora=no builtin_libvorbis=no builtin_libwebp=no builtin_mbedtls=no builtin_miniupnpc=no builtin_pcre2=no builtin_zlib=no builtin_zstd=no scu_build=all
```

**Comment by akien-mga** [comment]
Thanks for the contribution! I'll need to review this in depth. I appreciate the effort to preserve compatibility, but at the same time I'm not too fond of the added complexity and relative hackiness of what we need to do for this. But I'm not asking for any change of design, I need to review properly first and see if I eventually agree with the design decision, or want to propose an alternative approach.

**Comment by akien-mga** [comment]
That's fair! I'll probably revisit this myself later as I agree it would be a good change.

**Pull request closed without merging**